mlkem/ntt.c

```diff
@@ -9,137 +9,402 @@
- *         Montgomery form and signed canonical.
- * - start: Offset to the beginning of the butterfly block
- * - len: Index difference between coefficients subject to a butterfly
- * - bound: Ghost variable describing coefficient bound: Prior to `start`,
- *          coefficients must be bound by `bound + MLKEM_Q`. Post `start`,
- *          they must be bound by `bound`.
- * When this function returns, output coefficients in the index range
- * [start, start+2*len) have bound bumped to `bound + MLKEM_Q`.
- * Example:
- * - start=8, len=4
- *   This would compute the following four butterflies
- *          8     --    12
- *             9    --     13
- *                10   --     14
- *                   11   --     15
- * - start=4, len=2
- *   This would compute the following two butterflies
- *          4 -- 6
- *             5 -- 7
+
+/* 1.1 Forward NTT - Optimized C implementation
+ * --------------------------------------------
+ *
+ * Overview: this implementation aims to strike a balance
+ * between readability, formal verifification of type-safety,
+ * and performance.
+ *
+ * Layer merging
+ * -------------
+ * In this implementation:
+ *  Layers 1,2,3 and merged in function ntt_layer123()
+ *  Layers 4 and 5 are merged in function ntt_layer45()
+ *  Layer 6 stands alone in function ntt_layer6()
+ *  Layer 7 stands alone in function ntt_layer7()
+ *
+ * This particular merging was determined by experimentation
+ * and measurement of performance.
+ *
+ * Coefficient Reduction
+ * ---------------------
+ * The code defers reduction of coefficients in the core
+ * "butterfly" functions in each layer, so that the coefficients
+ * grow in absolute magnitiude by MLKEM_Q after each layer.
+ * This growth is reflected and verification in the contracts
+ * and loop invariants of each layer and their inner functions.
+ *
+ * Auto-Vectorization
+ * ------------------
+ * The code is written in a style that is amenable to auto-vectorization
+ * on targets that can support it, and using comtemporary compilers - for
+ * example, Graviton/AArch64 or Intel/AVX2 using recent versions of GCC
+ * or CLANG.
+ *
+ * Inner loops of the butterfly functions process at most 8 coefficients
+ * at once, since 8 * 16-bit = 128 bits, which is the width of a Vector
+ * register on many targets.
+ *
+ * Inlining of the inner "butterfly" functions is selectively enabled
+ * to allow a compiler to inline, to perform partial application, and
+ * vectorization of the top-level layer functions.
  */
-STATIC_TESTABLE
-void ntt_butterfly_block(int16_t r[MLKEM_N], int16_t zeta, int start, int len,
-                         int bound)
+
+#define NTT_BOUND1 (MLKEM_Q - 1)
+#define NTT_BOUND2 (2 * MLKEM_Q - 1)
+#define NTT_BOUND4 (4 * MLKEM_Q - 1)
+#define NTT_BOUND6 (6 * MLKEM_Q - 1)
+#define NTT_BOUND7 (7 * MLKEM_Q - 1)
+#define NTT_BOUND8 (8 * MLKEM_Q - 1)
+
+/* "pc" == "polynomial coefficients" */
+typedef int16_t pc[MLKEM_N];
```

Comment:
I find this type abbreviation rather confusing -- can we remove it?

Reply:
Removed.